// idylle/playground/actions/auth/login.js
const jwt = require('jsonwebtoken');
const sha1 = require('sha1');
const utils = require('../../utils');

module.exports = app => {
    const User = app.models.User;
    const salt = app.settings.security.salt;
    const ttl = app.settings.security.ttl;

    return Action({
        execute: (context) => {
            let email = context.data.email;
            let password = sha1(context.data.password);

            return findUser()
                .then(utils.ensureOne({code: 404, reason: 'user.not.found'}))
                .then(generateToken);


            function findUser() {
                return User.findOne({
                    email: email,
                    password: password
                })
            }

            function generateToken(user) {
                let token = {
                    user: user.toJSON()
                };

                return new Promise((resolve, reject) => {
                    return jwt.sign(token, salt, {expiresIn: ttl}, (err, encryptedToken) => {
                        if (err)
                            return reject(err);

                        return resolve(encryptedToken);
                    });
                });
            }
        }
    });
};
