// idylle/playground/actions/projects/collaborators.exclude.js
const Promise = require('bluebird');

module.exports = app => {
    const Project = app.models.Project;

    return Action({
        execute: context => {
            let project = Utils.closure();

            return findProject()
                .then(Utils.ensureOne({code: 422, reason: 'unprocessable.project'}))
                .then(project.assign)
                .then(ensureIssuerIsOwner)
                .then(ensureAllCollaborator)
                .then(removeFromProject)
                .then(removeFromCollaborators)
                .then(context.noContent);

            function findProject() {
                return Project.findById(context.params.id);
            }

            function ensureIssuerIsOwner(project) {
                return project.owner.toString() === context.user.id ? project : context.error('not.owner', 403)
            }

            function ensureAllCollaborator(project) {
                let collaborators = (context.data.collaborators instanceof Array) ? context.data.collaborators : [context.data.collaborators];
                return project.collaborators.some(sc => collaborators.some(c => sc.toString() !== c)) ? context.error('unprocessable.collaborators', 422) : true;
            }

            function removeFromProject() {
                project.value().collaborators.remove(context.data.collaborators);
                return project.value().save();
            }

            function removeFromCollaborators() {
                return Promise.each(context.data.collaborators || [], id => app.actions.users.update({
                    params: {
                        id: id
                    },
                    data: {
                        $pull: {
                            projects: project.value().id
                        }
                    }
                }))
            }
        }
    });
};
