// idylle/playground/models/Project.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const TimestampsPlugin = require('mongoose-timestamps');

const ProjectSchema = Schema({
    name: {
        type: String,
        required: true
    },
    client: {
        type: String,
        required: true
    },
    address: {
        street: {
            type: String,
            required: true
        },
        zip_code: {
            type: String,
            required: true
        },
        city: {
            type: String,
            required: true
        },
        state: {
            type: String
        },
        country: {
            type: String,
            required: true
        }
    },
    owner: {
        type: Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    collaborators: {
        type: [{
            type: Schema.Types.ObjectId,
            ref: 'User'
        }]
    },
    concretes: {
        type: [{
            type: Schema.Types.ObjectId,
            ref: 'Concrete'
        }]
    },
    castings: {
        type: [{
            type: Schema.Types.ObjectId,
            ref: 'Casting'
        }]
    },
    __v: {
        type: Number,
        select: false
    }
});

ProjectSchema.set('toJSON', {
    transform: (doc, result, options) => {
        result.id = result._id;
        delete result._id;
        return result;
    }
});

ProjectSchema.plugin(Utils.URIPlugin, {path: '/projects'});
ProjectSchema.plugin(TimestampsPlugin);

module.exports = mongoose.model('Project', ProjectSchema);
