// idylle/playground/models/User.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const TimestampsPlugin = require('mongoose-timestamps');

const UserSchema = Schema({
    first_name: {
        type: String
    },
    last_name: {
        type: String
    },
    email: {
        type: String,
        required: true,
        unique: true
    },
    password: {
        type: String,
        required: true,
        select: false
    },
    profile: {
        type: mongoose.Schema.ObjectId,
        ref: 'Profile'
    },
    __v: {
        type: Number,
        select: false
    },
    company: {
        type: String,
        required: true
    },
    projects: {
        type: [{
            type: Schema.Types.ObjectId,
            ref: 'Project'
        }]
    },
    preferences: {
        language: {
            type: String,
            default: 'fr'
        }
    }
});

UserSchema.set('toJSON', {
    transform: (doc, result, options) => {
        result.id = result._id;
        delete result._id;
        return result;
    }
});

UserSchema.plugin(Utils.URIPlugin, {path: '/users'});
UserSchema.plugin(TimestampsPlugin);

module.exports = mongoose.model('User', UserSchema);
