// idylle/playground/models/Casting.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const TimestampsPlugin = require('mongoose-timestamps');

const ConcreteSchema = Schema({
    title: {
        type: String,
        required: true
    },
    date: {
        type: Date,
        required: true
    },
    concrete: {
        type: Schema.Types.ObjectId,
        ref: 'Concrete',
        required: true,
    },
    telltales: [{
        duration: {
            type: Number,
            required: true
        },
        value: String
    }],
    project: {
        type: Schema.Types.ObjectId,
        ref: 'Project'
    },
    __v: {
        type: Number,
        select: false
    }
});

ConcreteSchema.set('toJSON', {
    transform: (doc, result, options) => {
        result.id = result._id;
        delete result._id;
        return result;
    }
});

ConcreteSchema.plugin(Utils.URIPlugin, {path: '/castings'});
ConcreteSchema.plugin(TimestampsPlugin);

module.exports = mongoose.model('Casting', ConcreteSchema);
